### PR TITLE
Initial implementation.

### PR DESCRIPTION
**Pull request by benwheeler**

### CODE REVIEW

**Comment by benwheeler** [comment]
A simple implementation of an HTTP sink for Flume. Currently there are no unit tests, and the integration tests are run against a separate Flume instance (the config to do this is not supplied).

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        Status status = null;
+
+        Channel ch = getChannel();
+        Transaction txn = ch.getTransaction();
+        txn.begin();
+        try {
+            Event event = ch.take();
+            sinkCounter.incrementEventDrainAttemptCount();
+
+            if (event != null && event.getBody().length > 0) {
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: Can event.getBody() be null?

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+                LOG.debug("Sending request : " + new String(event.getBody()));
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: toString or new String()

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                        LOG.error(String.format("Bad request, returned status code %s, event consumed", statusCode));
+
+                    } else if (statusCode == -1) {
+                        txn.rollback();
+                        status = Status.BACKOFF;
+
+                        LOG.debug("Malformed response returned from server, retrying");
+
+                    } else {
+                        txn.commit();
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: Do you want to commit for 5xx statuses here? Bad Gateway?

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                    httpClient.setRequestProperty("Content-Type", contentTypeHeader);
+                    httpClient.setRequestProperty("Accept", acceptHeader);
+                    httpClient.setConnectTimeout(connectTimeout);
+                    httpClient.setReadTimeout(requestTimeout);
+                    httpClient.setDoOutput(true);
+                    httpClient.setDoInput(true);
+                    httpClient.connect();
+
+                    OutputStream outputStream = httpClient.getOutputStream();
+                    outputStream.write(event.getBody());
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: If this write blows up, are streams always closed?

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                    outputStream.flush();
+                    outputStream.close();
+
+                    int statusCode = httpClient.getResponseCode();
+                    LOG.debug("Got status code : " + statusCode);
+
+                    httpClient.getInputStream().close();
+                    LOG.debug("Response processed and closed");
+
+                    if (statusCode == HttpURLConnection.HTTP_OK) {
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: From Dom: Don't have a github account but Datastream will currently return a 204 NoContent for a successful request.

File: src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+            LOG.error("Error sending HTTP request, retrying", t);
+
+            status = Status.BACKOFF;
+
+            // re-throw all Errors
+            if (t instanceof Error) {
+                throw (Error)t;
+            }
+
```
Comment (mnd999) [inline src/main/java/uk/gov/hmrc/flume/sink/HttpSink.java]: Might consider cleaning up the branching a little bit to reduce some duplication

**Pull request merged**